# Q1

1. Las top 10 fechas donde hay más tweets. Mencionar el usuario (username) que más publicaciones tiene por cada uno de esos días. Debe incluir las siguientes funciones:
```python
def q1_time(file_path: str) -> List[Tuple[datetime.date, str]]:
```
```python
def q1_memory(file_path: str) -> List[Tuple[datetime.date, str]]:
```
```python
Returns: 
[(datetime.date(1999, 11, 15), "LATAM321"), (datetime.date(1999, 7, 15), "LATAM_CHI"), ...]

Primero se hará esta parte: "Las top 10 fechas donde hay más tweets."

Solución:
La fecha viene en la columna date. Por lo tanto se debe hacer un groupby por fecha y contar la cantidad de tweets. Luego ordenar por cantidad descendentemente, y sacar las 10 primeras fechas, que serían las con más tweets.

In [55]:
import pandas as pd

path_tweets = 'farmers-protest-tweets-2021-2-4.json'
df_tweets_full = pd.read_json(path_tweets, lines = True)
print(f'Cantidad de filas: {len(df_tweets_full)}')
print(f'Cantidad de URLs unicas: {len(df_tweets_full["url"].unique())}')
df_tweets_full.dtypes

Cantidad de filas: 117407
Cantidad de URLs unicas: 117407


url                             object
date               datetime64[ns, UTC]
content                         object
renderedContent                 object
id                               int64
user                            object
outlinks                        object
tcooutlinks                     object
replyCount                       int64
retweetCount                     int64
likeCount                        int64
quoteCount                       int64
conversationId                   int64
lang                            object
source                          object
sourceUrl                       object
sourceLabel                     object
media                           object
retweetedTweet                 float64
quotedTweet                     object
mentionedUsers                  object
dtype: object

In [56]:
df_tweets_full.columns

Index(['url', 'date', 'content', 'renderedContent', 'id', 'user', 'outlinks',
       'tcooutlinks', 'replyCount', 'retweetCount', 'likeCount', 'quoteCount',
       'conversationId', 'lang', 'source', 'sourceUrl', 'sourceLabel', 'media',
       'retweetedTweet', 'quotedTweet', 'mentionedUsers'],
      dtype='object')

In [2]:
pd.options.display.max_colwidth = None
pd.options.display.max_columns = None
df_tweets_full.head(3)

url  \
0  https://twitter.com/ArjunSinghPanam/status/1364506249291784198   
1       https://twitter.com/PrdeepNain/status/1364506237451313155   
2   https://twitter.com/parmarmaninder/status/1364506195453767680   

                       date  \
0 2021-02-24 09:23:35+00:00   
1 2021-02-24 09:23:32+00:00   
2 2021-02-24 09:23:22+00:00   

                                                                                                                                                                                                                                                                                                             content  \
0                                              The world progresses while the Indian police and Govt are still trying to take India back to the horrific past through its tyranny. \n\n@narendramodi @DelhiPolice Shame on you. \n\n#ModiDontSellFarmers \n#FarmersProtest \n#FreeNodeepKaur https://t.co/es3kn0IQAF   
1  #FarmersProtest \n#ModiIgnoringFarmersDeaths \n#ModiDontSellFarmers \n@Kisanektamorcha \nFarmers constantly distroying crops throughout India. \nReally, it's hearts breaking...we care about our crops like our children. And govt. agriculture minister is laughing on us🚜🌾WE WILL WIN💪 https://t.co/kLspngG9xE   
2                                                                                                                                                                        ਪੈਟਰੋਲ ਦੀਆਂ ਕੀਮਤਾਂ ਨੂੰ ਮੱਦੇਨਜ਼ਰ ਰੱਖਦੇ ਹੋਏ \nਮੇਰੇ ਹਿਸਾਬ ਨਾਲ ਬਾਹਰ(ਪ੍ਰਦੇਸ਼) ਜਾਣ ਨਾਲੋਂ ਬਿਹਤਰ ਆ ਭਾਰਤ 'ਚ ਪੈਟਰੋਲ ਪੰਪ ਪਾ ਲਈਏ। 🤫🤫🤔🤔\n#FarmersProtest   

                                                                                                                                                                                                                                                                                                     renderedContent  \
0                                          The world progresses while the Indian police and Govt are still trying to take India back to the horrific past through its tyranny. \n\n@narendramodi @DelhiPolice Shame on you. \n\n#ModiDontSellFarmers \n#FarmersProtest \n#FreeNodeepKaur twitter.com/ravisinghka/st…   
1  #FarmersProtest \n#ModiIgnoringFarmersDeaths \n#ModiDontSellFarmers \n@Kisanektamorcha \nFarmers constantly distroying crops throughout India. \nReally, it's hearts breaking...we care about our crops like our children. And govt. agriculture minister is laughing on us🚜🌾WE WILL WIN💪 https://t.co/kLspngG9xE   
2                                                                                                                                                                        ਪੈਟਰੋਲ ਦੀਆਂ ਕੀਮਤਾਂ ਨੂੰ ਮੱਦੇਨਜ਼ਰ ਰੱਖਦੇ ਹੋਏ \nਮੇਰੇ ਹਿਸਾਬ ਨਾਲ ਬਾਹਰ(ਪ੍ਰਦੇਸ਼) ਜਾਣ ਨਾਲੋਂ ਬਿਹਤਰ ਆ ਭਾਰਤ 'ਚ ਪੈਟਰੋਲ ਪੰਪ ਪਾ ਲਈਏ। 🤫🤫🤔🤔\n#FarmersProtest   

                    id  \
0  1364506249291784198   
1  1364506237451313155   
2  1364506195453767680   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            user  \
0  {'username': 'ArjunSinghPanam', 'displayna

Se pasará la columna "date" a tipo date, es decir, se eliminara la parte horaria.

In [3]:
df_tweets_full['date_only'] = df_tweets_full['date'].dt.date
df_tweets_full[['date', 'date_only']]

date   date_only
0      2021-02-24 09:23:35+00:00  2021-02-24
1      2021-02-24 09:23:32+00:00  2021-02-24
2      2021-02-24 09:23:22+00:00  2021-02-24
3      2021-02-24 09:23:16+00:00  2021-02-24
4      2021-02-24 09:23:10+00:00  2021-02-24
...                          ...         ...
117402 2021-02-12 01:37:02+00:00  2021-02-12
117403 2021-02-12 01:36:53+00:00  2021-02-12
117404 2021-02-12 01:36:50+00:00  2021-02-12
117405 2021-02-12 01:36:49+00:00  2021-02-12
117406 2021-02-12 01:36:49+00:00  2021-02-12

[117407 rows x 2 columns]

In [4]:
cantidad_url_por_fecha = df_tweets_full.groupby('date_only')['url'].count().reset_index()
cantidad_url_por_fecha

date_only    url
0   2021-02-12  12347
1   2021-02-13  11296
2   2021-02-14  10249
3   2021-02-15   9197
4   2021-02-16  10443
5   2021-02-17  11087
6   2021-02-18   9625
7   2021-02-19   8204
8   2021-02-20   8502
9   2021-02-21   7532
10  2021-02-22   7071
11  2021-02-23   8417
12  2021-02-24   3437

In [5]:
top10_fechas = cantidad_url_por_fecha.rename(columns={'url':'count'}).sort_values(by='count', ascending=False).head(10)
top10_fechas

date_only  count
0   2021-02-12  12347
1   2021-02-13  11296
5   2021-02-17  11087
4   2021-02-16  10443
2   2021-02-14  10249
6   2021-02-18   9625
3   2021-02-15   9197
8   2021-02-20   8502
11  2021-02-23   8417
7   2021-02-19   8204

In [6]:
top_fechas = top10_fechas['date_only'].to_list()

In [7]:
top_fechas

[datetime.date(2021, 2, 12),
 datetime.date(2021, 2, 13),
 datetime.date(2021, 2, 17),
 datetime.date(2021, 2, 16),
 datetime.date(2021, 2, 14),
 datetime.date(2021, 2, 18),
 datetime.date(2021, 2, 15),
 datetime.date(2021, 2, 20),
 datetime.date(2021, 2, 23),
 datetime.date(2021, 2, 19)]

Ahora que se tienen las fechas, hay que buscarlas en el archivo original de tweets, ya que se quiere saber la cantidad de usuarios con mayor cantidad de tweets en esas fechas.

In [8]:
df_tweets_top_fechas = df_tweets_full[df_tweets_full['date_only'].isin(top_fechas)]
df_tweets_top_fechas.head(2)

url  \
3437  https://twitter.com/GaganSingh0095/status/1364364291609034760   
3438       https://twitter.com/harsimrin/status/1364364089154285570   

                          date  \
3437 2021-02-23 23:59:30+00:00   
3438 2021-02-23 23:58:42+00:00   

                                                                                                                                 content  \
3437                                                                                             @ClaudiaWebbe Thanks 🙏\n#FarmersProtest   
3438  #justicefornodeepkaur #FreeNodeepKaur #ReleaseNodeepKaur @amnesty @UN @humanrights1st @hrw #FarmersProtest https://t.co/iSbbPbMagV   

                                                                                                                             renderedContent  \
3437                                                                                                 @ClaudiaWebbe Thanks 🙏\n#FarmersProtest   
3438  #justicefornodeepkaur #FreeNodeepKaur #ReleaseNodeepKaur @amnesty @UN @humanrights1st @hrw #FarmersProtest twitter.com/someg1ll/statu…   

                       id  \
3437  1364364291609034760   
3438  1364364089154285570   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                user  \
3437                                                                                                                                                                                                                                                                                                                                                                  {'username': 'GaganSingh0095', 'displayname': 'BRANDED FARMER 🌾🚜🇮🇳', 'id': 1333210062123266050, 'description': '🇮🇳', 'rawDescription': '🇮🇳', 'descriptionUrls': [], 'verified': False, 'created': '2020-11-30T00:44:12+00:00', 'followersCount': 4, 'friendsCount': 22, 'statusesCount': 131, 'favouritesCount': 997, 'listedCount': 0, 'mediaCount': 83, 'location': '', 'protected': False, 'linkUrl': None, 'linkTcourl': None, 'profileImageUrl': 'https://pbs.twimg.com/profile_images/1348417164877852673/u6bgQlLZ_normal.jpg', 'profileBannerUrl': 'https://pbs.twimg.com/profile_banners/1333210062123266050/1613787092', 'url': 'https://twitter.com/GaganSingh0095'}   
3438  {'username': 'harsimrin', 'displayname': 'SainiG', 'id': 1358525631265447941, 'description': '26 y/o Mehndi Artist #kisaanmajdoorektazindabad #saveourfarmers #farmersprotest #istandwithfarmers Don’t take any tweets srs unless they srs. #4KidsorMore 😂🧡', 'rawDescription': '26 y/o Mehndi Artist #kisaanmajdoorektazindabad #saveourfarmers #farmersprotest #istandwithfarmers Don’t take any tweets srs unless they srs. #4KidsorMore 😂🧡', 'descriptionUrls': [], 'verified': False, 'created': '2021-02-07T21:18:56+00:00', 'followersCount': 122, 'friendsCount': 1797, 'statusesCount': 2096, 'favouritesCount': 2573, 'listedCount': 0, 'mediaCount': 42, 'location': 'Toronto ', 'protected': False, 'linkUrl': 'https://farmerprotests.carrd.co/', 'linkTcourl': 'https://t.co/E8ae

Aqui debo sacar el username de la columna user.

In [9]:
df_tweets_top_fechas['user']

3437                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          {'username': 'GaganSingh0095', 'displayname': 'BRANDED FARMER 🌾🚜🇮🇳', 'id': 1333210062123266050, 'description': '🇮🇳', 'rawDescription': '🇮🇳', 'descriptionUrls': [], 'verified': False, 'created': '2020-11-30T00:44:12+00:00', 'followersCount': 4, 'friendsCount': 22, 'statusesCount': 131, 'favouritesCount': 997, 'listedCount': 0, 'mediaCount': 83, 'location': '', 'protected': False, 'linkUrl': None, 'linkTcourl': None, 'profileImageUrl': 'https://pbs.twimg.com/profile_images/1348417164877852673/u6bgQlLZ_normal.jpg', 'pro

In [10]:
df_tweets_top_fechas['user'].iloc[0]

{'username': 'GaganSingh0095',
 'displayname': 'BRANDED FARMER 🌾🚜🇮🇳',
 'id': 1333210062123266050,
 'description': '🇮🇳',
 'rawDescription': '🇮🇳',
 'descriptionUrls': [],
 'verified': False,
 'created': '2020-11-30T00:44:12+00:00',
 'followersCount': 4,
 'friendsCount': 22,
 'statusesCount': 131,
 'favouritesCount': 997,
 'listedCount': 0,
 'mediaCount': 83,
 'location': '',
 'protected': False,
 'linkUrl': None,
 'linkTcourl': None,
 'profileImageUrl': 'https://pbs.twimg.com/profile_images/1348417164877852673/u6bgQlLZ_normal.jpg',
 'profileBannerUrl': 'https://pbs.twimg.com/profile_banners/1333210062123266050/1613787092',
 'url': 'https://twitter.com/GaganSingh0095'}

In [11]:
df_tweets_top_fechas['username'] = df_tweets_top_fechas['user'].apply(lambda x: x['username'])
df_tweets_top_fechas['username']

C:\Users\migue\AppData\Local\Temp\ipykernel_19800\1035973453.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_tweets_top_fechas['username'] = df_tweets_top_fechas['user'].apply(lambda x: x['username'])


3437      GaganSingh0095
3438           harsimrin
3439             JazzW54
3440       rkahlonsangha
3441      ik_stip_boeren
               ...      
117402     rickyrickstir
117403         PunjabTak
117404          ish_kayy
117405    TV9Bharatvarsh
117406         SikhVibes
Name: username, Length: 99367, dtype: object

Ahora que se tienen los users, se puede contar la cantidad de posts que hizo cada user en esas fechas.

In [12]:
cantidad_tweets_por_fecha_y_user = df_tweets_top_fechas.groupby(by=['date_only', 'username'], sort=False)['url'].count().reset_index().\
    rename(columns={'url': 'count'})
cantidad_tweets_por_fecha_y_user

date_only         username  count
0      2021-02-23   GaganSingh0095      1
1      2021-02-23        harsimrin      1
2      2021-02-23          JazzW54      2
3      2021-02-23    rkahlonsangha      2
4      2021-02-23   ik_stip_boeren     65
...           ...              ...    ...
44154  2021-02-12  DalwinderSaini5      1
44155  2021-02-12        ParamBal8      1
44156  2021-02-12        dissent07      1
44157  2021-02-12  Satveer22950341      1
44158  2021-02-12    rickyrickstir      1

[44159 rows x 3 columns]

In [13]:
mayor_cantidad_tweets_de_un_usuario_por_fecha = cantidad_tweets_por_fecha_y_user.\
    groupby(by=['date_only'], sort=False)['count'].agg(['max', 'sum']).reset_index()
mayor_cantidad_tweets_de_un_usuario_por_fecha

date_only  max    sum
0  2021-02-23  135   8417
1  2021-02-20  108   8502
2  2021-02-19  267   8204
3  2021-02-18  195   9625
4  2021-02-17  185  11087
5  2021-02-16  133  10443
6  2021-02-15  134   9197
7  2021-02-14  119  10249
8  2021-02-13  178  11296
9  2021-02-12  176  12347

In [14]:
join_dfs = cantidad_tweets_por_fecha_y_user.merge(mayor_cantidad_tweets_de_un_usuario_por_fecha, on='date_only', how='inner')
usuario_max_tweets_por_fecha = join_dfs[join_dfs['count'] == join_dfs['max']].reset_index(drop=True).sort_values(by='sum', ascending=False)
usuario_max_tweets_por_fecha

date_only         username  count  max    sum
9  2021-02-12  RanbirS00614606    176  176  12347
8  2021-02-13  MaanDee08215437    178  178  11296
4  2021-02-17   RaaJVinderkaur    185  185  11087
5  2021-02-16           jot__b    133  133  10443
7  2021-02-14    rebelpacifist    119  119  10249
3  2021-02-18  neetuanjle_nitu    195  195   9625
6  2021-02-15           jot__b    134  134   9197
1  2021-02-20  MangalJ23056160    108  108   8502
0  2021-02-23       Surrypuria    135  135   8417
2  2021-02-19         Preetm91    267  267   8204

In [15]:
lista_usuario_max_tweets_por_fecha = [(x, y) for x, y in usuario_max_tweets_por_fecha[['date_only', 'username']].values]
lista_usuario_max_tweets_por_fecha

[(datetime.date(2021, 2, 12), 'RanbirS00614606'),
 (datetime.date(2021, 2, 13), 'MaanDee08215437'),
 (datetime.date(2021, 2, 17), 'RaaJVinderkaur'),
 (datetime.date(2021, 2, 16), 'jot__b'),
 (datetime.date(2021, 2, 14), 'rebelpacifist'),
 (datetime.date(2021, 2, 18), 'neetuanjle_nitu'),
 (datetime.date(2021, 2, 15), 'jot__b'),
 (datetime.date(2021, 2, 20), 'MangalJ23056160'),
 (datetime.date(2021, 2, 23), 'Surrypuria'),
 (datetime.date(2021, 2, 19), 'Preetm91')]

# Profiling

In [4]:
from q1_time import q1_time
import cProfile
import pstats

file_path = "farmers-protest-tweets-2021-2-4.json"
with cProfile.Profile() as profile:
    result_q1_time = q1_time(file_path)

results = pstats.Stats(profile)
results.sort_stats(pstats.SortKey.CUMULATIVE)
results.print_stats()

c:\Users\migue\Documents\Búsqueda de Trabajo\Entrevistas\Latam\challenge_DE_01\src\q1_time.py:14: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_tweets_top_fechas['username'] = df_tweets_top_fechas['user'].apply(lambda x: x['username'])


         718121 function calls (717709 primitive calls) in 18.141 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.006    0.006   18.141   18.141 c:\Users\migue\Documents\Búsqueda de Trabajo\Entrevistas\Latam\challenge_DE_01\src\q1_time.py:5(q1_time)
    12/10    0.031    0.003   17.524    1.752 c:\Users\migue\AppData\Local\Programs\Python\Python310\lib\site-packages\pandas\util\_decorators.py:302(wrapper)
        1    0.000    0.000   17.508   17.508 c:\Users\migue\AppData\Local\Programs\Python\Python310\lib\site-packages\pandas\util\_decorators.py:166(wrapper)
        1    0.140    0.140   17.477   17.477 c:\Users\migue\AppData\Local\Programs\Python\Python310\lib\site-packages\pandas\io\json\_json.py:315(read_json)
        1    0.044    0.044   15.299   15.299 c:\Users\migue\AppData\Local\Programs\Python\Python310\lib\site-packages\pandas\io\json\_json.py:730(read)
        1    0.000    0.000   13.123   13

In [5]:
results.sort_stats(pstats.SortKey.LINE)
results.print_stats()

         718121 function calls (717709 primitive calls) in 18.141 seconds

   Ordered by: line number

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
       88    0.000    0.000    0.000    0.000 {built-in method __new__ of type object at 0x00007FFAB114A920}
      127    0.000    0.000    0.000    0.000 {built-in method _abc._abc_instancecheck}
        1    0.000    0.000    0.000    0.000 {built-in method _abc._abc_subclasscheck}
        1    0.000    0.000    0.000    0.000 {built-in method _codecs.lookup}
        1    0.283    0.283    0.283    0.283 {built-in method _codecs.utf_8_decode}
        1    0.000    0.000    0.000    0.000 {built-in method _codecs.lookup_error}
        1    0.000    0.000    0.000    0.000 {method 'append' of 'collections.deque' objects}
        6    0.000    0.000    0.000    0.000 {built-in method _locale.setlocale}
       69    0.000    0.000    0.000    0.000 {method 'match' of 're.Pattern' objects}
       61    0.000    0.000

Se ve que la función toma en total 23.108 segundos, siendo 22.313  segundos solo de la lectura del json (read_json), el resto de métodos no parecen afectar tanto el tiempo de ejecución. Por lo que lo primordial es disminuir el tiempo de lectura del json.

In [11]:
%load_ext memory_profiler

In [16]:
%%mprun -f  q1_time
q1_time(file_path)

c:\Users\migue\Documents\Búsqueda de Trabajo\Entrevistas\Latam\challenge_DE_01\src\q1_time.py:14: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_tweets_top_fechas['username'] = df_tweets_top_fechas['user'].apply(lambda x: x['username'])


Filename: c:\Users\migue\Documents\Búsqueda de Trabajo\Entrevistas\Latam\challenge_DE_01\src\q1_time.py

Line #    Mem usage    Increment  Occurrences   Line Contents
     5    159.6 MiB    159.6 MiB           1   def q1_time(file_path: str) -> List[Tuple[datetime.date, str]]:
     6   1641.1 MiB   1481.4 MiB           1       df_tweets_full = pd.read_json(file_path, lines = True)
     7   1646.4 MiB      5.4 MiB           1       df_tweets_full['date_only'] = df_tweets_full['date'].dt.date
     8                                         
     9   1647.7 MiB      1.3 MiB           1       cantidad_url_por_fecha = df_tweets_full.groupby('date_only')['url'].count().reset_index()
    10   1647.7 MiB      0.0 MiB           1       top10_fechas = cantidad_url_por_fecha.rename(columns={'url':'count'}).sort_values(by='count', ascending=False).head(10)
    11   1647.7 MiB      0.0 MiB           1       top_fechas = top10_fechas['date_only'].to_list()
    12                                      

La memoria que usa es 1676 MiB, siendo 1481 MiB solo por leer el JSON. Otras sumas importantes de memoria ocurren cuando se genera la columna date_only, que suma 5.1 MiB; y cuando se genera el dataframe df_tweets_top_Fechas que suma 19 MiB.